evm_loader/program/src/solana_backend.rs

```diff
@@ -14,105 +14,42 @@ use solana_sdk::{
 };
 use std::{
     cell::RefCell,
+    cell::Ref,
 };
 
-use crate::solidity_account::SolidityAccount;
-use crate::account_data::AccountData;
 use solana_sdk::program::invoke;
 use solana_sdk::program::invoke_signed;
 use std::convert::TryInto;
 use arrayref::{array_ref, array_refs, array_mut_ref, mut_array_refs};
 
-fn keccak256_digest(data: &[u8]) -> H256 {
-    H256::from_slice(Keccak256::digest(&data).as_slice())
-}
-
 pub fn solidity_address<'a>(key: &Pubkey) -> H160 {
     H256::from_slice(key.as_ref()).into()
 }
 
-fn u256_to_h256(value: U256) -> H256 {
-    let mut v = vec![0u8; 32];
-    value.to_big_endian(&mut v);
-    H256::from_slice(&v)
+pub trait AccountStorage {
+    fn contract_id(&self) -> H160;
+    fn get_account_solana_address(&self, address: H160) -> Option<&Pubkey>;
+    fn get_contract_seeds(&self) -> Option<(H160, u8)>;
+    fn get_caller_seeds(&self) -> Option<(H160, u8)>;
+    fn exists(&self, address: H160) -> bool;
+    fn basic(&self, address: H160) -> Basic;
+    fn code_hash(&self, address: H160) -> H256;
+    fn code_size(&self, address: H160) -> usize;
+    fn code(&self, address: H160) -> Vec<u8>;
+    fn storage(&self, address: H160, index: H256) -> H256;
+    fn block_number(&self) -> U256;
+    fn block_timestamp(&self) -> U256;
 }
 
-pub struct SolanaBackend<'a> {
-    accounts: Vec<Option<SolidityAccount<'a>>>,
-    aliases: RefCell<Vec<(H160, usize)>>,
-    clock_account: &'a AccountInfo<'a>,
-    account_infos: &'a [AccountInfo<'a>],
+pub struct SolanaBackend<'a, 's> {
+    account_storage: Ref<'s ,dyn AccountStorage>,
```

Comment:
Why do you need to hold `account_storage` as Ref?

Reply:
Actually i couldn't find better solution to pass AccountStorage to SolanaBackend without borrowing errors